File: foothold_app/android/app/src/main/java/dev/rohit/foothold/capture/CaptureConfig.java
package dev.rohit.foothold.capture;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.HashSet;
import java.util.Set;

/**
 * On-device config + the tagged name→number registry (the privacy boundary).
 * Synced from GET /api/v1/companies (dm_name → last-10 digits), so a WhatsApp
 * notification whose title is a tagged lead's name (incl. Hindi) resolves to
 * that lead's number. Untagged chats never leave the device.
 */
public class CaptureConfig {
    private final SharedPreferences p;

    public CaptureConfig(Context ctx) {
        p = ctx.getSharedPreferences("foothold_capture", Context.MODE_PRIVATE);
    }

    public String baseUrl() { return p.getString("baseUrl", "https://foothold-yantrai.fly.dev"); }
    public void setBaseUrl(String v) { p.edit().putString("baseUrl", v.replaceAll("/+$", "")).apply(); }

    public String token() { return p.getString("token", ""); }
    public void setToken(String v) { p.edit().putString("token", v.trim()).apply(); }

    public Set<String> taggedNames() { return p.getStringSet("taggedNames", new HashSet<String>()); }
    public void setTaggedNames(Set<String> s) { p.edit().putStringSet("taggedNames", s).apply(); }

    public String lastKey() { return p.getString("lastKey", ""); }
    public void setLastKey(String v) { p.edit().putString("lastKey", v).apply(); }

    /** last-10 digits of a phone, or null. */
    public static String digits(String phone) {
        if (phone == null) return null;
        String d = phone.replaceAll("[^0-9]", "");
        return d.length() >= 10 ? d.substring(d.length() - 10) : null;
    }

    /** Resolve a WhatsApp notification title (contact display name, incl. Hindi)
     *  to a tagged lead number. Case-insensitive exact match on the name. */
    public String digitsForName(String title) {
        if (title == null) return null;
        String t = title.trim().toLowerCase();
        for (String pair : taggedNames()) {
            int i = pair.indexOf("||");
            if (i > 0 && pair.substring(0, i).trim().toLowerCase().equals(t)) {
                return pair.substring(i + 2);
            }
        }
        return null;
    }
}

File: foothold_app/android/capture/CaptureConfig.kt
package dev.rohit.foothold.capture

import android.content.Context

/**
 * On-device config + the tagged-number registry for Foothold capture.
 *
 * Auth to the Foothold backend is a BEARER TOKEN (the same FOOTHOLD_TOKEN the
 * server uses for /api/v1). The captured touch is matched to a lead server-side
 * by phone (last-10 digits) — see targets/db.py::find_lead_by_phone.
 *
 * The tagged set is the privacy boundary: capture services only ever process
 * numbers that appear here (people saved as leads). Everyone else is dropped on
 * the device before anything leaves it ("tagged-contacts-only").
 *
 * ⚠ WRITTEN, NOT DEVICE-VERIFIED in this environment — build + test on a Mac
 *   with adb (see capture/README.md).
 */
class CaptureConfig(ctx: Context) {
    private val p = ctx.getSharedPreferences("foothold_capture", Context.MODE_PRIVATE)

    var baseUrl: String
        get() = p.getString("baseUrl", "https://foothold-yantrai.fly.dev")
            ?: "https://foothold-yantrai.fly.dev"
        set(v) = p.edit().putString("baseUrl", v.trimEnd('/')).apply()

    /** FOOTHOLD_TOKEN — sent as `Authorization: Bearer <token>`. */
    var token: String
        get() = p.getString("token", "") ?: ""
        set(v) = p.edit().putString("token", v.trim()).apply()

    /** last call-log row timestamp we already processed (dedupe) */
    var lastCallTs: Long
        get() = p.getLong("lastCallTs", 0L)
        set(v) = p.edit().putLong("lastCallTs", v).apply()

    /** tagged numbers, stored as last-10-digits */
    var taggedDigits: Set<String>
        get() = p.getStringSet("taggedDigits", emptySet()) ?: emptySet()
        set(v) = p.edit().putStringSet("taggedDigits", v).apply()

    /** "name||digits" pairs so the WhatsApp listener can resolve a contact name → number */
    var taggedNamePairs: Set<String>
        get() = p.getStringSet("taggedNames", emptySet()) ?: emptySet()
        set(v) = p.edit().putStringSet("taggedNames", v).apply()

    fun isTagged(phone: String?): Boolean = digits(phone)?.let { taggedDigits.contains(it) } ?: false

    /** resolve a WhatsApp notification title (a contact/display name) to a tagged number */
    fun digitsForName(title: String?): String? {
        if (title.isNullOrBlank()) return null
        val t = title.trim().lowercase()
        return taggedNamePairs.firstOrNull { it.substringBefore("||").lowercase() == t }
            ?.substringAfter("||")
    }

    companion object {
        fun digits(phone: String?): String? {
            val d = phone?.filter { it.isDigit() } ?: return null
            return if (d.length >= 10) d.takeLast(10) else null
        }
    }
}
